lib/commons/dom/find-up.js

```diff
@@ -1,34 +1,61 @@
 /* global dom, axe */
+dom.oldFindUp = function (element, target) {
+	let matches,
+		parent = element;
+	// Phantom and IE 8
+	do {// recursively walk up the DOM, checking each parent node
+		parent = parent.parentNode;
+		if (!matches) {
+			matches = document.querySelectorAll(target);
+			matches = axe.utils.toArray(matches);
+			if (!matches.length) {
+				return null;
+			}
+		}
+	} while (parent && !matches.includes(parent));
+
+	return parent;
+};
+
 /**
  * recusively walk up the DOM, checking for a node which matches a selector
  *
  * **WARNING:** this should be used sparingly, as it's not even close to being performant
  * @method findUp
  * @memberof axe.commons.dom
  * @instance
- * @param {HTMLElement|String} element The starting HTMLElement
+ * @param {HTMLElement|VirtualNode} element The starting HTMLElement or its virtualNode equivalent
  * @param {String} target The selector for the HTMLElement
  * @return {HTMLElement|null} Either the matching HTMLElement or `null` if there was no match
  */
 dom.findUp = function (element, target) {
-	let doc, matches,
-		parent = element;
+	let parent;
 
-	do {// recursively walk up the DOM, checking each parent node
-		parent = (parent.assignedSlot ? parent.assignedSlot : parent.parentNode);
-		if (parent && parent.nodeType === 11) {
-			matches = null;
-			parent = parent.host;
+	if (typeof element.actualNode === 'undefined') {
+		element = axe.utils.getNodeFromTree(axe._tree[0], element);
+	}
+	parent = element.actualNode;
+	if (!element.shadowId && typeof element.actualNode.closest === 'function') {
+		let match = element.actualNode.closest(target);
+		if (match) {
+			return match;
 		}
-		if (!matches) {
-			doc = axe.commons.dom.getRootNode(parent);
-			matches = doc.querySelectorAll(target);
-			matches = axe.utils.toArray(matches);
-			if (doc === document && !matches.length) {
-				return null;
+		return null;
+	}
+	// handle shadow DOM nodes and older browsers
+	if (typeof parent.matches === 'function') {
+		do {// recursively walk up the DOM, checking each parent node
+			parent = (parent.assignedSlot ? parent.assignedSlot : parent.parentNode);
```

Comment:
no, we don't want to use getComposedParent because a rule may have to look for the `<slot>` itself (for example)

Not sure I understand what you mean by "match the `contents` element"?